### DISCUSSION TITLE
Eclipse Kura 6.0.0 Release Plan

### DISCUSSION THREAD

**Discussion by MMaiero** [eclipse-kura/kura#5410] [General]
In progress document on the features and goals for Eclipse Kura 6.0.0.

This major release will bring:

- OSGi update to latest release to leverage the latest improvements and compatibility with newer java versions
- Old Networking removal (breaking change)
- New Networking improvements for feature parity with old networking
- Eclipse Kura Modularisation process: The scope is to investigate and start the process of modularisation. The modularisation base idea would be to have a minimal core of Eclipse Kura with networking, firewall and container management. All the rest would be applications that could be distributed in separate form. One would be, for example, wires. Another could be the cloud connection to Eclipse Kapua, etc.
  - Externalise ESF Web UI: externalise the web UI is the main modularisation split and target for this release. We need to define architecturally how to split the framework, keeping aligned the constraints and prove the modularisation approach by first splitting the Web UI and distributing it separately as an optional component. The target is to implement a first modularisation example. Choose one or more subsystems to extract and externalize as a separate component.
- Eclipse Kura Web UI - containers as first class citizens: Containers promoted to services from the current device tab positioning.

The target release date is end of Q1 2025 with milestone releases available from early Q1 2025

**Comment by alexcorvis84** [comment]
Hi @MMaiero,

I'm concern about point listed: Old Networking removal (breaking change). 

Does it mean that in future release 6.0.0 networking is going to change completely as it is now running in <6.0.0 versions? I mean, its configuration based on NetworkManager and NetworkConfiguration/FirewallConfigurationService config from snapshot_0.xml will change? Can you share some specific details what's in the roadmap about this breaking change?

Many thanks in advance.

**Comment by MMaiero** [comment]
Hi,
thanks for the message. No, there is no breaking change for the snapshot. The work we have done with the networkmanager provider relies on the existing (documented) properties for the snapshot and firewall.
You can have already a look at the documentation and properties. If a certain property is not marked as deprecated, then it will not be touched or removed.
From the snapshot point of view, this change will have no or minimal impact.

Best regards

**Reply by alexcorvis84** [reply]
Hi @MMaiero,
I've been checking the [documentation](http://eclipse-kura.github.io/kura/docs-release-5.6/gateway-configuration/network-configuration/#ipv4-properties) and I can´t find any property marked as deprecated. So I'm not sure if the source I'm looking as the right one...

Apart from that, I have noticed in the source code that several [network templates](https://github.com/eclipse-kura/kura/tree/develop/kura/distrib/src/main/resources/common/snapshots) have been created. I think I understand that the new way of generating the snapshot_0.xml now depends on some [option that is passed](https://github.com/eclipse-kura/kura/blob/develop/kura/distrib/src/main/resources/generic-arm32/kura_install.sh#L93C1-L93C86) to add the template dynamically at compiling time. Am I right?

Thanks in advance.

**Reply by pierantoniomerlino** [reply]
Hi @alexcorvis84,
regarding the network template files, they are used to customize Kura configuration **during the installation**. The installer detects the network interfaces of the device and creates a snapshot specific for the device. Look [here](https://eclipse-kura.github.io/kura/docs-develop/getting-started/install-kura/#initial-network-configuration) for more details.

The option passed to the customize script is used only for selection a standard or NN profile (that does not require the properties for the networking).

**Reply by alexcorvis84** [reply]
Yeah, just as thought. 

Just to note: quite interesting that form version >5.5.0 there's a need to execute the function **[setup_libudev](https://github.com/eclipse-kura/kura/blob/develop/kura/distrib/src/main/resources/common/customize-installation.sh#L15)** to make some amendments to make it retrocompatible. I guess this is a _temporal_ fix? Do you know if this will be solved in a cleaner way in the future from source devs?

I mention this, because I still _try to maintain_ the scripts that generate an upgrade script from one version to the next, and I had to add this function to run in the [extract_upgrade.sh](https://github.com/eclipse-kura/kura/blob/develop/kura/distrib/src/main/sh/extract_upgrade.sh) script, because otherwise Kura was not able to start after upgrading a device from 5.4.0 to 5.5.0 with the script generated.

PS: the reason I still rely on this _old but gold_ method is quite simple, we've devices running on 4G communications, and size matters for SIM data consumption. Installer is bigger than updater, which is half the size of the file :)

Many thanks for replying!

**Reply by mattdibi** [reply]
Hello there @alexcorvis84,

> Just to note: quite interesting that form version >5.5.0 there's a need to execute the function [setup_libudev](https://github.com/eclipse-kura/kura/blob/develop/kura/distrib/src/main/resources/common/customize-installation.sh#L15) to make some amendments to make it retrocompatible. I guess this is a temporal fix? Do you know if this will be solved in a cleaner way in the future from source devs?

There's no plan currently to change this behaviour. We would happily accept a contribution on this matter though.

Keep in mind that, officially and by design, we don't support upgrades from one version of kura to the other.

**Reply by alexcorvis84** [reply]
Thanks for replying @mattdibi 

I know that upgrade scripts were left unsupported since 4.1.1 version... but (for the reason exposed) with some tweaks I got them working again and I use them with every new version.

I'd happily make a contribution for the libudev fix, but as you [linked](https://github.com/eclipse-kura/kura/blob/KURA_5.5.0_RELEASE/kura/distrib/src/main/resources/common/customize-installation.sh#L17) into script, seems it's the best way to fix actually 😅

Regards

**Comment by alexcorvis84** [comment]
Hi @mattdibi @MMaiero ,

I'm just wondering what's the status of Kura 6.0 RC as I've no seen no updates since original post.

**Reply by mattdibi** [reply]
Hi @alexcorvis84,

thanks for reaching out! We're still working on it but we're close to being feature complete. Our initial plan might have been a little too optimistic. Probably @MMaiero can share more details about the current timeline.
